### DISCUSSION TITLE
Global configuration for default store and, config, and script

### DISCUSSION THREAD

**Discussion by wdkrnls** [ropensci/targets#583] [Ideas]
Hi,

I would like to build a wrapper around my project for targets. For that I want to save workflows in a non standard location. Right now I have to write wrappers over every targets function to specify those locations. I'd rather just specify those once to a package global configuration and then have targets "just work". Would such a feature be welcome in targets?

**Comment by wlandau** [comment]
`tar_config_set()` and `tar_config_get()` interact with a project-level `_targets.yaml` file with configuration settings that include the store (usually `_targets/`) and script (usually `_targets.R`). `tar_config_set()` and `tar_config_get()` each support a `config` argument to allow you to work with YAML config files at paths other than `_targets.yaml`. `tar_config_get()` with the default `config = _targets.yaml` supplies many default arguments to functions like `tar_read()`. That's about as far as I can go with global configuration. `_targets.yaml` is already a configuration file, and I am avoiding configuration through the environment/options because those settings go away when you restart your R session (been there before with `drake`).

**Reply by wdkrnls** [reply]
That was all clear to me from your excellent documentation. I guess I'll stick with my wrapper approach, then.

**Reply by stuvet** [reply]
I'm doing something similar for a couple of my targets to keep assets within Google Cloud storage buckets while still using `aws_qs` (the file metadata gets passed as the target output). 

This limits my 'external' (to GCP) network bandwidth charges & keeps my scratch file size & memory usage on my main node under control.

For another purpose @wlandau suggested `tarchetype` hooks in #581. I suspect the wrapper code could be condensed into a `tarchetypes::tar_hook_outer` & applied to selected targets to do what you suggest @wdkrnls 

@wlandau I suspect such side effects would be strongly discouraged, but could this be another solution for this use case with appropriately severe warnings?

**Reply by wdkrnls** [reply]
Thanks for the suggestion. Looking over the `tar_hook_outer` example, however, makes me think that maybe my terminology wasn't very clear. I don't want to preprocess each target for a particular script/workflow. I want to make a convention that helps me keep organized and in control of all my workflows. For that I was hoping that the default value for the config yaml file location could be changed from it's own default project *relative* path to another *absolute* path of my choosing. In pseudocode the functions `tar_config_get` would go from

```r
function(..., config = '_targets.yaml') ...
```
to become

```r
function (..., config = get('default_config_yaml_path', tar_globals))
```
The same kind of default arguments would exist for `tar_config_set` and all the functions which specify `script` and `store` paths.

**Reply by wlandau** [reply]
Yeah, hooks seem like a different situation.

What specifically is appealing about the ability to switch the default `config` during a session? Typically, projects have independent file systems, and you can switch to the project by changing the working directory to the root of that project. RStudio project files make managing multiple projects really convenient if you are using the RStudio IDE, and VSCode has similar file/project management tools.

**Reply by stuvet** [reply]
Sorry I misunderstood - much clearer after reading the title properly!

**Reply by wdkrnls** [reply]
I find projects to be a bit heavy interface. I have gotten into the habit of using them for package development (and usually can manage 4-5 of them open at a time: usually where a workflow depends on a package which in turn depends on another package...), but for many small data analysis projects dealing with their setup and  managing the IDE state (which then changes the global environment) it feels like projects require a superhuman amount of discipline to me. I prefer to offload discipline to R as much as possible. I also want to get as many of my fellow researchers using targets as possible, and their level of tolerance for workflows requiring discipline on computer (vs in the lab) is much lower than mine.

**Reply by wdkrnls** [reply]
Mostly, people are happy with creating many Rmarkdowns within a project. So I was hoping to use e.g. target markdown with a standardized separate location for where the workflows themselves get stored. That seemed like a closer to ideal trade off in human complexity to me.

**Reply by wlandau** [reply]
I think I see what you mean now. If each Target Markdown report has a different pipeline, you could use `root.dir` to make sure each one gets its own working sub-directory. Maybe use a setup chunk like this one at the top of each report:

    ```{r setup}
    fs::dir_create("subdirectory_of_this_report")
    knitr::opts_knit$set(root.dir = "subdirectory_of_this_report")
    ```

Somewhat related: multiple Target Markdown reports can share the same pipeline, e.g. a whole `bookdown` project for a single pipeline.

**Reply by wlandau** [reply]
So I just changed my mind on some of this, and development `targets` now has multi-project config via `_targets.yaml`: https://books.ropensci.org/targets/projects.html#multiple-projects. That feature set includes environment variables `TAR_CONFIG` (to set the path to the YAML file to something other than `_targets.yaml`) and `TAR_PROJECT` (to set the project within the YAML file).
